### PR TITLE
feat(profiling): add support for pytorch profiling

### PR DESCRIPTION
**Pull request by sanchda**
PR does
- Patches `torch.profiler.profile` class by adding our own `on_trace_ready` handler
- Adds GPU time/flops/memory samples via libdatadog interface in `on_trace_ready` event handler
- Ensures that libdd exporter is enabled if pytorch is enabled
- Hides functionality behind a FF set to False by default
- changelog entry
- Is there a minimum python version?
  - the biggest requirement is that the current pytorch profiler API which we instrument was introduced in torch version 1.8.1 (https://pytorch.org/blog/introducing-pytorch-profiler-the-new-and-improved-performance-tool/),  do we just want to document or we could disable the instrumentation if we detect an outdated version  with `torch.__version__`
- Some documentation on needed user configuration, conflicting features, gotchas

~~Probably should make experimental/beta collectors not part of the ALL template (Is this blocking since we haven't done in the past??)~~

## Testing Done
- Tested by running on ec2 GPU instance
- Tested by running `prof-pytorch` service in staging
- I'm not entirely sure if we need unit tests for this feature, or where they would live. Would we want the unit test suite to depend on torch? Maybe this is solved for tracing integrations, though

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/profiling/collector/pytorch.py [hunk 1]
```diff
@@ -90,6 +82,11 @@ def _set_original(
     def _start_service(self):
         # type: (...) -> None
         """Start collecting framework profiler usage."""
+        try:
+            import torch
```
Comment (peterg17) [inline ddtrace/profiling/collector/pytorch.py:86]: good extra safety check!

**Review by peterg17 — COMMENTED** [review]
Thanks!

File: ddtrace/profiling/collector/pytorch.py
Comment (taegyunkim) [inline ddtrace/profiling/collector/pytorch.py:1]: I was thinking about how we can write tests for this, and realized that we don't have CUDA gpus for testing locally. 

Hmm, maybe it's possible to write one using Mac GPUs, and run only on Apple Silicone Macs? 

Circle CI has GPU execution environments, https://circleci.com/docs/using-gpu/, but not sure whether they're available for our repo and I personally find it easier to iterate when I can run tests locally.

**Review by taegyunkim — CHANGES_REQUESTED** [review]

File: ddtrace/profiling/collector/pytorch.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    ):
+        # type: (...) -> None
+        self._torch_module.profiler.profile = value
+
+
+def handle_torch_trace(prof):
+    NANOS_PER_MICROSECOND = 1e3
+    LOG.debug("handle_torch_trace called")
+    # need an upper bound of events collected, can be adjusted based on profile size
+    single_profile_event_limit = 1_000_000
```
[the remark this answers is not in the document]
Comment (peterg17) [reply ddtrace/profiling/collector/pytorch.py:127]: good idea

File: ddtrace/profiling/collector/pytorch.py
Comment (peterg17) [reply ddtrace/profiling/collector/pytorch.py:1]: > Hmm, maybe it's possible to write one using Mac GPUs, and run only on Apple Silicone Macs?

Unfortunately the PyTorch profiler doesn't support metal gpu profiling, though PyTorch itself can be executed on Metal (maybe a fun hackathon project to contribute this 😄 ). Hm thanks for the link! Do you know where the circle CI config is in the repo?

Comment (taegyunkim) [reply ddtrace/profiling/collector/pytorch.py:1]: I am not the expert in editing the circle ci config, but my guess is that
- you need to add a section in the .circleci/config.templ.yml file for pytorch https://github.com/DataDog/dd-trace-py/pull/9461/files#diff-8166c31d8ba62357b4a7cac4407d8744e60cf6b6ba32b09af939807db4c8f490
which will be referring to a venv defined in riotfile.py 
- riotfile.py would have https://github.com/DataDog/dd-trace-py/pull/9461/files#diff-8c2dff2cbcc0f4fff63050381d4c7546ae7195383ec8f9cc993f8494b77a34c3

File: ddtrace/profiling/collector/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    num_events_collected = min(len(prof.events()), single_profile_event_limit)
+    trace_start_us = prof.profiler.kineto_results.trace_start_us()
+    for i, e in enumerate(prof.events()[:num_events_collected]):
```
[the remark this answers is not in the document]
Comment (peterg17) [reply ddtrace/profiling/collector/pytorch.py]: good point, fixed

File: ddtrace/profiling/collector/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        device_name = "cuda " + str(e.device_index)
+        end_time_us = int(trace_start_us + e.time_range.end)
+        event_duration_us = e.time_range.elapsed_us()
```
[the remark this answers is not in the document]
Comment (peterg17) [reply ddtrace/profiling/collector/pytorch.py]: fixed, thanks!

File: ddtrace/profiling/collector/pytorch.py
Comment (peterg17) [reply ddtrace/profiling/collector/pytorch.py:1]: Discussed offline, might be worth pursuing different approaches to testing on a gpu machine. We do have a service running internally that tests this, which provides some integration testing.

File: ddtrace/profiling/collector/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        super(MLProfilerCollector, self)._start_service()
+
+    def _stop_service(self):
+        # type: (...) -> None
+        """Stop collecting framework profiler usage."""
+        super(MLProfilerCollector, self)._stop_service()
+        self.unpatch()
+
+    def patch(self):
+        # type: (...) -> None
```
Comment (emmettbutler) [inline ddtrace/profiling/collector/pytorch.py:74]:
```suggestion
    def patch(self) -> None :
```

**Review by emmettbutler — APPROVED** [review]

**Review by taegyunkim — DISMISSED** [review]

File: ddtrace/profiling/collector/pytorch.py
Comment (peterg17) [reply ddtrace/profiling/collector/pytorch.py:1]: Basically after talking with @danielsn  we do need some way of testing this on GPU in an open-source way, and need to evaluate the best way of doing this

**Review by ZStriker19 — APPROVED** [review]
From API side looks good

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/profiling/collector/pytorch.py (hunk 5 above)
Comment (danielsn) [reply ddtrace/profiling/collector/pytorch.py:74]: What's the issue with the type annotation here?

Comment (danielsn) [reply ddtrace/profiling/collector/pytorch.py:74]: i.e. should I apply this everywhere in this file?

File: ddtrace/profiling/collector/pytorch.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        recorder: Recorder,
+        tracer: typing.Optional[Tracer],
+    ) -> None:
+        wrapt.ObjectProxy.__init__(self, wrapped)
+        self.on_trace_ready = handle_torch_trace
+        self._self_recorder = recorder
+        self._self_tracer = tracer
+
+
+@attr.s
```
[the remark this answers is not in the document]
Comment (danielsn) [reply ddtrace/profiling/collector/pytorch.py]: done

File: ddtrace/profiling/collector/pytorch.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    ):
+        # type: (...) -> None
+        self._torch_module.profiler.profile = value
+
+
+def handle_torch_trace(prof):
+    NANOS_PER_MICROSECOND = 1e3
+    LOG.debug("handle_torch_trace called")
+    # need an upper bound of events collected, can be adjusted based on profile size
+    num_events_collected = min(len(prof.events()), pytorch_events_limit)
```
[the remark this answers is not in the document]
Comment (danielsn) [reply ddtrace/profiling/collector/pytorch.py:115]: Inlined the events_limits (we've got a default value which we're setting using `or`, LMK if there is a better way to do that"

File: .github/workflows/pytorch_gpu_tests.yml [hunk 8]
```diff
@@ -0,0 +1,40 @@
+name: Pytorch Unit Tests (with GPU) 
+
+on:
+  pull_request:
+    branches:
+      - 'main'
```
Comment (taegyunkim) [inline .github/workflows/pytorch_gpu_tests.yml:6]:
```suggestion
      - 'main'
      - 'mq-working-branch**'
```

I think this is needed to run this GH action for merge queue

File: .github/workflows/pytorch_gpu_tests.yml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    paths:
+      - 'ddtrace/profiling/collector/pytorch.py'
```
Comment (taegyunkim) [inline .github/workflows/pytorch_gpu_tests.yml:8]:
```suggestion
  push:
    branches:
  	  - 'main'
  	  - 'mq-working-branch**'
  pull_request:
    paths:
      - 'ddtrace/profiling/collector/pytorch.py'
```

I think this should let us run this for main and mergequeue pushes, and for pull requests modifying the pytorch.py file

File: .github/workflows/pytorch_gpu_tests.yml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+   unit-tests:
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        arch: [x86_64]
+    runs-on: ${{ matrix.os }} 
+    steps:
+      - uses: actions/checkout@v4
+        # Include all history and tags
+        with:
```
Comment (taegyunkim) [inline .github/workflows/pytorch_gpu_tests.yml:21]:
```suggestion
        with:
          persist-credentials: false
```

as in [ci: do not persiste git credentials in GitHub Actions](https://github.com/DataDog/dd-trace-py/pull/11640/files)

File: .github/workflows/pytorch_gpu_tests.yml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        name: Install Python
+        with:
+          python-version: '3.12'
+
+      - uses: actions-rust-lang/setup-rust-toolchain@v1
+      - name: Install latest stable toolchain and rustfmt
+        run: rustup update stable && rustup default stable && rustup component add rustfmt clippy
+
+      - name: Install hatch
+        run: pip install hatch
```
Comment (taegyunkim) [inline .github/workflows/pytorch_gpu_tests.yml:34]:
```suggestion
      - name: Install hatch
        uses: pypa/hatch@install
        with:
          version: "1.12.0"
```

File: .github/workflows/pytorch_gpu_tests.yml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+      
+      - name: Install PyTorch
+        run: pip install torch
```
Comment (taegyunkim) [inline .github/workflows/pytorch_gpu_tests.yml:37]: Would you mind pinning the version? or you could specify range of versions in the hatch.toml file

File: releasenotes/notes/profiling-add-pytorch-integration-0683123b7bb83f99.yaml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    profiling: Adds an experimental integration with the PyTorch profiler which can be enabled
+    by setting ``DD_PROFILING_PYTORCH_ENABLED=true``. This feature instruments the PyTorch 
+    profiler API (https://pytorch.org/docs/stable/_modules/torch/profiler/profiler.html)
+    so that GPU profiling data can be sent to Datadog for visualization.
+    This feature requires the environment variable ``DD_PROFILING_EXPORT_LIBDD_ENABLED=true``
```
Comment (taegyunkim) [inline releasenotes/notes/profiling-add-pytorch-integration-0683123b7bb83f99.yaml:8]: Should we say 'This feature also enables profiling native exporter' instead? Users don't need to set `DD_PROFILING_EXPORT_LIBD_ENABLED`. We will just auto enable it.

File: tests/profiling/test_main.py [hunk 14]
```diff
@@ -37,6 +39,28 @@ def test_call_script_gevent(monkeypatch):
...
+def test_call_script_pytorch_gpu(tmp_path, monkeypatch):
+    filename = str(tmp_path / "pprof")
+    monkeypatch.setenv("DD_PROFILING_OUTPUT_PPROF", filename)
+    monkeypatch.setenv("DD_PROFILING_ENABLED", "1")
+    monkeypatch.setenv("DD_PROFILING_PYTORCH_ENABLED", "1")
+    stdout, stderr, exitcode, pid = call_program(
+        "ddtrace-run", sys.executable, os.path.join(os.path.dirname(__file__), "simple_program_pytorch_gpu.py")
+    )
+
+    output_pprof_filename = filename + "." + str(pid) + ".1"
```
[the remark this answers is not in the document]
Comment (peterg17) [reply tests/profiling/test_main.py]: removed this bit of code now

File: .github/workflows/pytorch_gpu_tests.yml (hunk 12 above)
Comment (peterg17) [reply .github/workflows/pytorch_gpu_tests.yml:40]: ok, i'll start by pinning a recent working version `2.5.1`

Comment (peterg17) [reply .github/workflows/pytorch_gpu_tests.yml:40]: Nvm we should support any torch version >=1.8.1

File: ddtrace/profiling/collector/pytorch.py (hunk 5 above)
Comment (danielsn) [reply ddtrace/profiling/collector/pytorch.py:74]: https://github.com/DataDog/dd-trace-py/issues/11729 to track

File: releasenotes/notes/profiling-add-pytorch-integration-0683123b7bb83f99.yaml (hunk 13 above)
Comment (peterg17) [reply releasenotes/notes/profiling-add-pytorch-integration-0683123b7bb83f99.yaml]: sounds good, I removed mention of the native exporter here

**Review by danielsn — APPROVED** [review]
